Docs for cluster tiering to improve query concurrency (#12128)
apache/druid: docs/operations/mixed-workloads.md

This situation slows down subsequent queries—heavy and light—and may trigger timeout errors for those queries.
With proper resource isolation, you can execute long-running, low priority queries without interfering with short-running, high priority queries.

Druid provides the following strategies to isolate resources and improve query concurrency:
- **Query laning** where you set a limit on the maximum number of long-running queries executed on each Broker. 
- **Service tiering** which defines separate groups of Historicals and Brokers to receive different query assignments based on query priority.

You can profile Druid queries using normal performance profiling techniques such as Druid query metrics analysis, thread dumps of JVM, or flame graphs to identify what resources are affected by mixed workloads.
The largest bottleneck will likely be in the Broker HTTP threads.
Mitigate resource contention of the Broker HTTP threads with query laning.
However, mixed workloads also affect other resources, including processing threads and merge buffers.
Reduce the burden on these resources by applying service tiering in addition to query laning.

## Query laning

> Query laning directs Druid to restrict resource usage for less urgent queries to ensure dedicated resources for high priority queries. Query laning is ideal when you need to run many concurrent queries having heterogeneous workloads.

Author: ```suggestion
When you need to run many concurrent queries having heterogeneous workloads, start with query laning to optimize your query performance.
Query laning restricts resource usage for less urgent queries to ensure dedicated resources for high priority queries.
```

* `druid.query.scheduler.laning.isLimitPercent` – Whether to treat the lane limit as an exact number or a percent of the minimum of `druid.server.http.numThreads` or `druid.query.scheduler.numThreads`.

With manual laning, incoming queries can be labeled with the desired lane in the `lane` parameter of the [query context](../querying/query-context.md).

See [Query prioritization and laning](../configuration/index.md#query-prioritization-and-laning) for additional details on query laning configuration.

### Example

Example config for query laning with the high/low laning strategy:

```
# Laning strategy
druid.query.scheduler.laning.strategy=hilo
druid.query.scheduler.laning.maxLowPercent=20

# Limit the number of HTTP threads for query processing
# This value should be less than druid.server.http.numThreads
druid.query.scheduler.numThreads=40
```

## Service tiering

In service tiering, you define separate groups of Historicals and Brokers to manage queries based on the segments and resource requirements of the query.
You can limit the resources that are set aside for certain types of queries.

> Many heavy queries involving complex subqueries or large result sets can crash a Broker, or worse, take down a cluster.

Reviewer: This sounds a bit scarier than necessary as it technically can happen but is not common. Can we mention the problem of resource contention between long-running queries and short-running queries instead?

Author: ```suggestion
Many heavy queries involving complex subqueries or large result sets can hog resources away from high priority, interactive queries.
```

Minimize the impact of these heavy queries by limiting them to a separate Broker tier.

> When all Brokers set aside for heavy queries are occupied, users must wait to submit additional heavy queries until the designated resources become available.

Reviewer: This looks a bit unclear to me. I think it's because there is nothing in Druid disallowing users to submit additional queries. Perhaps 

```suggestion
When all Brokers set aside for heavy queries are occupied, additional queries submitted subsequently can wait until the designated resources become available, which can fail those queries after query timeout expires.
```

Author: ```suggestion
When all Brokers set aside for heavy queries are occupied, subsequent queries must wait until the designated resources become available.
A prolonged wait results in the later queries failing with a timeout error.
```

---
id: mixed-workloads
title: Configure Druid for mixed workloads
sidebar_label: Mixed workloads
---

<!--
  ~ Licensed to the Apache Software Foundation (ASF) under one
  ~ or more contributor license agreements.  See the NOTICE file
  ~ distributed with this work for additional information
  ~ regarding copyright ownership.  The ASF licenses this file
  ~ to you under the Apache License, Version 2.0 (the
  ~ "License"); you may not use this file except in compliance
  ~ with the License.  You may obtain a copy of the License at
  ~
  ~   http://www.apache.org/licenses/LICENSE-2.0
  ~
  ~ Unless required by applicable law or agreed to in writing,
  ~ software distributed under the License is distributed on an
  ~ "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
  ~ KIND, either express or implied.  See the License for the
  ~ specific language governing permissions and limitations
  ~ under the License.
  -->

If you frequently run concurrent, heterogeneous workloads on your Apache Druid cluster, configure Druid to properly allocate cluster resources to optimize your overall query performance.

> Each Druid query consumes a certain amount of cluster resources, such as processing threads, memory buffers for intermediate query results, and HTTP threads for communicating between Brokers and data servers.

Reviewer 2: Great job describing  the "heavy" vs "light" here.
nit: Do we need the carriage return at the end of line 28?

Author: the single carriage return leaves the content in the same paragraph so I like to have it to make it easier to track future changes

For example, if you run both sets of queries in the same Druid cluster, heavy queries may employ all available HTTP threads.
This situation slows down subsequent queries—heavy and light—and may trigger timeout errors for those queries.
With proper resource isolation, you can execute long-running, low priority queries without interfering with short-running, high priority queries.

Druid provides the following strategies to isolate resources and improve query concurrency:
- **Query laning** where you set a limit on the maximum number of long-running queries executed on each Broker. 
- **Service tiering** which defines separate groups of Historicals and Brokers to receive different query assignments based on query priority.

You can profile Druid queries using normal performance profiling techniques such as Druid query metrics analysis, thread dumps of JVM, or flame graphs to identify what resources are affected by mixed workloads.
The largest bottleneck will likely be in the Broker HTTP threads.
Mitigate resource contention of the Broker HTTP threads with query laning.
However, mixed workloads also affect other resources, including processing threads and merge buffers.

> Reduce the burden on these resources by applying service tiering in addition to query laning.

Reviewer 2: ```suggestion
To reduce the burden on these resources, apply both service tiering and to query laning.
```
nit
Also not sure about the carriage returns in this 'graf

## Query laning

When you need to run many concurrent queries having heterogeneous workloads, start with query laning to optimize your query performance.
Query laning restricts resource usage for less urgent queries to ensure dedicated resources for high priority queries.

> Query lanes are analogous to carpool and normal lanes on the freeway. With query laning, Druid sets apart VIP lanes from other general lanes.

Reviewer 2: Wonder if there is another term than VIP (non-native speakers may trip over that one)

Author: ```suggestion
Query lanes are analogous to carpool and normal lanes on the freeway. With query laning, Druid sets apart particular lanes from other general lanes.
```

druid.query.scheduler.numThreads=40
```

## Service tiering

In service tiering, you define separate groups of Historicals and Brokers to manage queries based on the segments and resource requirements of the query.
You can limit the resources that are set aside for certain types of queries.
Many heavy queries involving complex subqueries or large result sets can hog resources away from high priority, interactive queries.
Minimize the impact of these heavy queries by limiting them to a separate Broker tier.
When all Brokers set aside for heavy queries are occupied, subsequent heavy queries must wait until the designated resources become available.
A prolonged wait results in the later queries failing with a timeout error.

Note that you can separate Historical processes into tiers without having separate Broker tiers.
Historical-only tiering is not sufficient to meet the demands of mixed workloads on a Druid cluster.
However, it is useful when you query certain segments more frequently than others, such as often analyzing the most recent data.
Historical tiering assigns data from specific time intervals to specific tiers in order to support higher concurrency on hot data. 

> The examples below demonstrate two tiers—hot and cold—for both the Historicals and Brokers. The Brokers will serve short-running, light queries before long-running, heavy queries. Light queries will be routed to the hot tiers, and heavy queries will be routed to the cold tiers.

Reviewer 2: I notice that we switch to hot and cold here. Are these related to "heavy" and "light" introduced above?

Reviewer 2: Not for this PR, but I would like to see an example of how  two queries, one heavy/cold or light/hot and how they would be routed using this strategy. I think that might be a good result of the research of using the flame graphs, etc.

Author: In this doc I'm using "hot/cold" for data and tiers and "heavy/light" for describing the queries. The crucial bit that connects the two concepts is in L111: "Light queries will be routed to the hot tiers, and heavy queries will be routed to the cold tiers." Is that clear enough?

Reviewer 2: Sounds good.

The largest bottleneck will likely be in the Broker HTTP threads.
Mitigate resource contention of the Broker HTTP threads with query laning.
However, mixed workloads also affect other resources, including processing threads and merge buffers.

> To reduce the burden on these resources, apply both service tiering and to query laning.

Author: ```suggestion
To reduce the burden on these resources, apply both service tiering and query laning.
```
